RFC.md

```diff
@@ -105,345 +63,485 @@ Thus:
+Signature: CMS signed data, according to RFC5652
+- certificate of user
+- Signed data:
+    - create byte array of JSON binary and decrypted metadata key
+    - SignerIdentifier of the CMS container contains userId of the user who created the signature
+
+The encoded binary CMS structure is base64-encoded and sent in the header of the metadata request.
+The Section [Uploading payload and metadata](#uploading-payload-and-metadata-file-step-5--8) further describes how the signature is handled.
+
+### Verifying the metadata
+Before using the metadata file, e.g. on a folder refresh the client has to verify the signature.
+The client has to find its metadata-key in the `recipients` list and decrypt it.
+Then the client passes the JSON binary and the decrypted metadata-key to the verification algorithm.
+
+If any of the following checks fail, the client needs to refuse further sync and informs the user:
+- check counter: new counter must be greater than locally stored counter
+- check signature
+- check that hash of metadata-key is in keyChecksums
+- check that no hash has been removed from the keyChecksums
+- check if cert of all recipients is issued by the CA
+
+### Modifying and accessing content of an end-to-end encrypted folder
+In general, clients need to perform two steps to modify the content of an end-to-end encrypted folder.
+Firstly, clients upload, modify or delete the actual files via the WebDAV API and secondly modify the metadata JSON accordingly.
+
+#### Modifying content of end-to-end encrypted folders
+The following steps are required to create, update, delete files of an end-to-end encrypted folder.
+If any of the following steps failed, show a verbose warning to the user, unlock the folder and abort.
+
+1. Make a backup of file to be modified and metadata file, so that in case of any upload problems the client can revert to a consistent state.
+2. Detect topmost folder which needs to be changed
+  1. Increase counter of the metadata of this folder
+  2. Lock the folder with increased counter passed as header X-NC-E2EE-COUNTER
+3. Check for changes in the encrypted folder.
+  1. if current version is up to date, go to next step 4.
+  2. If not current,
+    - get the latest metadata file
+    - verify metadata file
+    - decrypt metadata file
+    - check if file operation is still possible (e.g. file to be renamed can already be deleted)
+4. Perform specific steps to create/update/delete file or folder:
+    * Create new file:
+      1. Generate a new 128-bit encryption key for the file and encrypt it using AES/GCM/NoPadding
+      2. Generate a new 96-bit IV for the file
+      3. Encrypt the file using the key and IV with AES/GCM/NoPadding
+      4. Create new random identifier by generating a random UUID and removing the dash (`-`). The identifier must follow `/^[0-9a-fA-F]{32}$/` TODO: use a better id, or ignore "-"
+      5. Add new file to the files array in the metadata file with UUID as identifier and 
```

Comment:
The way this is designed right now you have to encrypt each file in one go. That's pretty bad when handling larger files. You should probably add the option to encrypt in chunks (stream). So this will need a block size, which, if it's not global, should probably go in here.